staking/parse.go

```diff
@@ -3,13 +3,161 @@
+func ParseCertificatePermissive(bytes []byte) (*Certificate, error) {
+	if len(bytes) > MaxCertificateLen {
+		return nil, ErrCertificateTooLarge
+	}
+
+	input := cryptobyte.String(bytes)
+	// Consume the length and tag bytes.
+	if !input.ReadASN1(&input, cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedCertificate
+	}
+
+	// Read the "to be signed" certificate into input.
+	if !input.ReadASN1(&input, cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedTBSCertificate
+	}
+	if !input.SkipOptionalASN1(cryptobyte_asn1.Tag(0).Constructed().ContextSpecific()) {
+		return nil, ErrMalformedVersion
+	}
+	if !input.SkipASN1(cryptobyte_asn1.INTEGER) {
+		return nil, ErrMalformedSerialNumber
+	}
+	if !input.SkipASN1(cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedSignatureAlgorithmIdentifier
+	}
+	if !input.SkipASN1(cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedIssuer
+	}
+	if !input.SkipASN1(cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedValidity
+	}
+	if !input.SkipASN1(cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedIssuer
+	}
+
+	// Read the "subject public key info" into input.
+	if !input.ReadASN1(&input, cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedSPKI
+	}
+
+	// Read the public key algorithm identifier.
+	var pkAISeq cryptobyte.String
+	if !input.ReadASN1(&pkAISeq, cryptobyte_asn1.SEQUENCE) {
+		return nil, ErrMalformedPublicKeyAlgorithmIdentifier
+	}
+	var pkAI asn1.ObjectIdentifier
+	if !pkAISeq.ReadASN1ObjectIdentifier(&pkAI) {
+		return nil, ErrMalformedOID
+	}
+
+	// Note: Unlike the x509 package, we require parsing the public key.
+
+	var spk asn1.BitString
+	if !input.ReadASN1BitString(&spk) {
+		return nil, ErrMalformedSubjectPublicKey
+	}
+	publicKey, signatureAlgorithm, err := parsePublicKey(pkAI, spk)
+	return &Certificate{
+		Raw:                bytes,
+		SignatureAlgorithm: signatureAlgorithm,
+		PublicKey:          publicKey,
+	}, err
+}
+
+// Ref: https://github.com/golang/go/blob/go1.19.12/src/crypto/x509/parser.go#L215-L306
+func parsePublicKey(oid asn1.ObjectIdentifier, publicKey asn1.BitString) (any, x509.SignatureAlgorithm, error) {
+	der := cryptobyte.String(publicKey.RightAlign())
+	switch {
+	case oid.Equal(oidPublicKeyRSA):
+		pub := &rsa.PublicKey{N: new(big.Int)}
+		if !der.ReadASN1(&der, cryptobyte_asn1.SEQUENCE) {
+			return nil, 0, ErrInvalidRSAPublicKey
+		}
+		if !der.ReadASN1Integer(pub.N) {
+			return nil, 0, ErrInvalidRSAModulus
+		}
+		if !der.ReadASN1Integer(&pub.E) {
+			return nil, 0, ErrInvalidRSAPublicExponent
+		}
+
+		if pub.N.Sign() <= 0 {
+			return nil, 0, ErrRSAModulusNotPositive
+		}
+		// Ref: https://github.com/golang/go/blob/go1.19.12/src/crypto/tls/handshake_client.go#L874-L877
+		if bitLen := pub.N.BitLen(); bitLen > MaxRSAKeyBitLen {
+			return nil, 0, fmt.Errorf("%w: bitLen=%d > maxBitLen=%d", ErrInvalidRSAPublicKey, bitLen, MaxRSAKeyBitLen)
+		}
+		if pub.E <= 0 {
+			return nil, 0, ErrRSAPublicExponentNotPositive
+		}
```

Comment:
I would restrict N and E further. I can't promise crypto/rsa will always accept E=1, or 256-bit keys, for example

Ideally, if your in-use certificates allow, you'd restrict N's bitsize to two or three options, and E to two or three values. If you need to be more flexible than that, I would check at least 3 <= E <= MaxInt32, and 2048 <= N.BitLen <= 8192.

Reply:
Oh, and also I'd check that both N and E are odd. (You can do that with `N.Bit(0)`)

N needs to be a product of two large primes, and E needs to be coprime with phi(N) which is even. Libraries can and do reject even values because some math tricks don't work on them.